### PR TITLE
Route to kubevirt VMs using infra id as service label selector

### PR DESCRIPTION
**Pull request by davidvossel**
As we're looking to expand the kubevirt platform to manage VMs on external infra (an ocp cluster external to the one hypershift runs on), it's possible that VMs from multiple clusters will get placed in the same namespace. As a result, we need to start labeling the components we place in these namespaces using the unique HostedCluster.Spec.InfraID. 

This serves two purposes for us.
1. We can ensure our ingress passthrough services only route to VMs within a specific cluster
2. We can clearly identify what resources associated with a HostedCluster need to be cleaned up on the external cluster (routes, services, VMs, etc...)

### CODE REVIEW

**Comment by davidvossel** [comment]
/hold

I want to see the kubevirt e2e lane pass before we consider merging this.

**Comment by davidvossel** [comment]
/cc @orenc1 maybe you can review this? What I've done is add a new label to the VMI pods that allows us to identify them by the Hosted CLuster (using infraID) that the VMs belong to. This will let us create Services and Network Policies that target a specific cluster within a namespace rather than all VMs in a namespace...

I think this level of identification for vms/vmis/pods could be important to us with your external infra work. It could be possible for people to use the same namespace on an external infra cluster to host VMs from multiple guest clusters.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go
Comment (orenc1) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go:132]: the string `"hypershift.kubevirt.io/infra-id"` appears in 2 locations. wouldn't it be better to put it in a variable, e.g. `infraIDLabelKey`, as you did in `MachineTemplateSpec` function

**Review by orenc1 — APPROVED** [review]
Thanks, it makes total sense to me to distinguish the ingress service to target the VMs that belong to the relevant guest cluster only.

i have just one nit, other than that, lgtm

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go
Comment (davidvossel) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/ingress/reconcile.go:132]: oh, yeah definitely. I didn't realize i hadn't transitioned to a variable for that. good call.

**Comment by orenc1** [comment]
/hold

i've now figured out we have another problem if multiple ingress services and routes are created in the same infra namespace for multiple guest clusters - the service and the route are created with an hardcoded name: `default-ingress-passthrough-service` and `default-ingress-passthrough-route`. in that case, subsequent ingress service and route will overwrite existing ones, making existing guest clusters unreachable.

**Comment by davidvossel** [comment]
this PR needs more work. We identified a few more area's the infra id needs to be applied

1. make our label a const
2. add the label so it's applied to LBs and PVCs mirrored into the guest as well

**Comment by orenc1** [comment]
> /hold cancel
> 
> @orenc1 I updated this PR. We now add the infra id to both the infra LBs and PVCs in order to help associate those resources with the HostedCluster.
> 
> I had to make a change to the cloud-provider-kubevirt config here [1] for this to work for the LBs. It doesn't matter what order these PRs get merged in though. Hypershift will attempt to set the labels on the cloud config, but they won't get actually applied to the LBs until [1] is merged. It doesn't break anything though.
> 
>     1. [Add id custom labels to infra Load balancers kubevirt/cloud-provider-kubevirt#206](https://github.com/kubevirt/cloud-provider-kubevirt/pull/206)

@davidvossel , looks like the ingress service and route still have a constant name:
service:
https://github.com/openshift/hypershift/blob/c6f74edb0a606d9e753b0c6dc6fd429f7920ff79/control-plane-operator/hostedclusterconfigoperator/controllers/resources/manifests/ingress.go#L41

route:
https://github.com/openshift/hypershift/blob/c6f74edb0a606d9e753b0c6dc6fd429f7920ff79/control-plane-operator/hostedclusterconfigoperator/controllers/resources/manifests/ingress.go#L50

if we're utilizing the same infra namespace for multiple guest clusters, we must have a unique service and route for each.

**Comment by davidvossel** [comment]
> if we're utilizing the same infra namespace for multiple guest clusters, we must have a unique service and route for each.

ugh... you're right. I missed this.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go
Comment (davidvossel) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:788]: @orenc1 i fixed the issue with the route and service not being unique by adding some unique generated strings to the end of these resources.

**Comment by davidvossel** [comment]
/hold

i just want to look at this a little closer

**Comment by davidvossel** [comment]
/hold cancel

I'm comfortable with this pr now.

**Comment by orenc1** [comment]
me too.
i'll lgtm once the ci lane passes.

File: api/v1alpha1/hostedcluster_types.go
Comment (orenc1) [inline api/v1alpha1/hostedcluster_types.go:683]: what does it mean "_is required once set_" ?
if GenerateID is required, it must not be empty or missing in the first place.
so there is no meaning to "once set" in my opinion.

and same for the validation message below - "is immutable once set" - if that field can't be empty when the resource is created, the value is set on creation, and the "once set" is superfluousץ

Comment (davidvossel) [reply api/v1alpha1/hostedcluster_types.go:683]: here's an explanation, and it's where i took the example from. https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/#immutablility-after-first-modification 

The idea is that I want GenerateID to be allowed to be set exactly once, and never changed after that. Including removing the value entirely.

The line you reference `!has(oldSelf.generateID) || has(self.generateID)` is kind of odd. I believe what it's saying is that it's okay for the value to be set only when the old value was empty.

**Comment by davidvossel** [comment]
/hold

i think this is good to go, but i want to run the new api field by another hypershift approver first

**Comment by davidvossel** [comment]
/hold

i talked with hypershift maintainers, we're good to go with this one.

**Review by orenc1 — APPROVED** [review]
/lgtm

**Review by orenc1 — APPROVED** [review]
/approve
/lgtm
/thanks

**Pull request merged**
